core/app/models/spree/promotion_code/batch_builder.rb

```diff
@@ -0,0 +1,66 @@
+class ::Spree::PromotionCode::BatchBuilder
+  attr_reader :promotion_code_batch
+  delegate :promotion, :number_of_codes, :base_code, to: :promotion_code_batch
+
+  class_attribute :random_code_length, :batch_size, :sample_characters
+  self.random_code_length = 6
+  self.batch_size = 1_000
+  self.sample_characters = ('a'..'z').to_a + (2..9).to_a.map(&:to_s)
+
+  def initialize(promotion_code_batch)
+    @promotion_code_batch = promotion_code_batch
+  end
+
+  def build_promotion_codes
+    generate_random_codes
+
+    Spree::PromotionCodeBatchMailer
+      .promotion_code_batch_finished(promotion_code_batch)
+      .deliver_now
+  rescue => e
+    promotion_code_batch.update!(error: e.inspect)
+    Spree::PromotionCodeBatchMailer
+      .promotion_code_batch_errored(promotion_code_batch)
+      .deliver_now
+    raise e
+  end
+
+  private
+
+  def generate_random_codes
+    total_batches = (number_of_codes.to_f / self.class.batch_size).ceil
+
+    total_batches.times do |i|
+      codes_for_current_batch = Set.new
+      codes_to_generate = [self.class.batch_size, number_of_codes - i * batch_size].min
+
+      while codes_for_current_batch.size < codes_to_generate
+        new_codes = Array.new(codes_to_generate) { generate_random_code }.to_set
+        codes_for_current_batch += get_unique_codes(new_codes)
+      end
+
+      codes_for_current_batch.map do |value|
+        promotion.codes.build(value: value, promotion_code_batch: promotion_code_batch)
+      end
+
+      promotion.save!
+
+      # We have to reload the promotion because otherwise all promotion codes
+      # we are creating will remain in memory. Doing a reload will remove all
+      # codes from memory.
+      promotion.reload
+    end
+  end
+
+  def generate_random_code
+    suffix = Array.new(self.class.random_code_length) do
+      sample_characters.sample
+    end.join
+
+    "#{base_code}_#{suffix}"
```

Comment:
This concatenation can not be configured anywhere, right? We should make this configurable, as I already tickets in my backlog ("remove the underscore form promotion codes") 😆